feature/builds

marcstreeter opened the pull request:
added builds faking endpoint. I tried to vary the output with all the different status', some output like the `project.yml` I made very short because I don't think the UI does anything with it. I could copy paste and example though if that's desired.

MiloCasagrande commented on src/routes/builds.js:
```diff
@@ -0,0 +1,297 @@
+const router = express.Router();
+
+router.get('/', (req, res) => {
+    res.status(200).json({
```
`res.status(200)` is not necessary if the status is really 200, just `res.json` is fine.

MiloCasagrande commented on src/routes/builds.js:
```diff
@@ -0,0 +1,297 @@
+    return;
+});
+
+export default router;
```
Missing a newline at the end of the file.

It's not mentioned in the README, but there is a linter command you can run to lint the code and apply a uniform style.

If you run `npm run lint` you'll lint the entire project, or you can lint just the files you changed with `npx prettier file1 file1 --write`.

MiloCasagrande commented on src/routes/builds.js:
```diff
@@ -0,0 +1,297 @@
+                        status: 'CANCELLING',
+                    },
+                ],
+                runs_url: `http://fakerunsurl/7`,
```
Backticks here are not necessary, just `'http://fakerunsurl/7'`. Or did you want to pass the build number as a parameter?

Maybe we can look into the [faker](https://www.npmjs.com/package/faker) package and its APIs and see if we can create fake URLs instead of `http://fakrunsurl`. Maybe we can use `https://example.net` as the base URL (as the de facto standard in example URLs).

MiloCasagrande commented on src/routes/builds.js:
```diff
@@ -0,0 +1,297 @@
+            build: {
+                // required
+                build_id: req.param.build,
+                url: 'http://fakebuildid/1',
```
In this case maybe we want to have the same `build` number in the URL.

MiloCasagrande commented on src/index.js:
```diff
@@ -44,6 +45,7 @@ app.use((req, res, next) => {
   next();
 });
 
+app.use('/projects/jobserv/builds', BuildsRoutes);
```
We have to extend this, it's too limited.

We might want to have a router that manages the `/projects` routes, and inside that we should have handlers for all the cases we need.

Consider a factory named `test-factory`: we need to be able to perform request with that factory name as the build parameter, like `/projects/test-factory`. 

This is how we will be able to access the factory builds in the same way as we will do with the real API.
Then, we need to be able to access `/projects/test-factory/lmp/builds` as that is the base path of our builds.

We should probably have handlers for:
```
GET /projects
GET /projects/:project
GET /projects/:project/builds
GET /projects/:project/lmp/builds
```

The `:project` parameter can be a factory name or something else (as in the `jobserv` example).
As a real example consider your `marcstreeter` factory in production: you have a `marcstreeter/lmp/builds` endpoint to get all your builds.

MiloCasagrande commented on src/routes/builds.js:
```diff
@@ -0,0 +1,297 @@
+        '5': 'RUNNING_WITH_FAILURES',
+        '6': 'RUNNING',
+        '7': 'RUNNING',
+    }[req.param.build];
```
I'd say we randomize the status value, not based on the build number.
If we want to look at build number 300, that will not work.

MiloCasagrande commented on src/routes/builds.js:
```diff
@@ -0,0 +1,297 @@
+    res.status(200).json({
+        status: 'success',
+        data: {
+            builds: [
```
The `builds` array needs to be randomized and automatically created. We might want to randomly return between 0 and $max (like 60, just because I used 60 for the number of devices returned) builds, or stick to a maximum number always.

You can take a look at the [devices.js](https://github.com/foundriesio/api-faker/blob/master/src/routes/devices.js) router and see how the `faker` package has been used to fake almost all the returned values.

We also have to keep in mind about pagination, but that can be dealt with later.

MiloCasagrande commented on src/routes/builds.js:
```diff
@@ -0,0 +1,297 @@
+                    build_id: 1,
+                    url: 'http://fakebuildid/1',
+                    status: 'FAILED',
+                    runs: [
```
Same here as for the `builds` array, this needs to be randomized and automatically generated.

MiloCasagrande commented on src/routes/builds.js:
```diff
@@ -0,0 +1,297 @@
+    return;
+});
+router.get('/:build/project.yml', (req, res) => {
+    res.status(200).type('text/yaml').send('some: yaml');
```
Better to have a "really fake" yaml file based on one of the actual yaml file.

doanac commented on src/routes/builds.js:
```diff
@@ -0,0 +1,297 @@
+            ],
+            total: 7,
+            next: 'https://fake/projects/jobserv/builds/?page=1&limit=10',
+        }
```
this is sort of doing pagination. I assume that's great cause the UI needs to handle that. However, wouldn't this response sort of make the UI go into an endless loop since it will always include the "next" link?

MiloCasagrande replied:
Pagination is necessary so we can do it on the UI, but the UI does not use the `next` link in the response. The UI relies on the server-side to retrieve the data (in a Backed-For-Frontend fashion, we don't do CORS), and the server part will build the `next` URL based on the data it holds.

MiloCasagrande commented on src/routes/projects.js:
```diff
@@ -0,0 +1,175 @@
+    status: 'success',
+    data: {
+      builds: builds,
+      total: builds.length,
```
I think we need to define a maximum for `total`, and make it fixed. `total` should be the total number of builds, not the number of builds returned.

The issue is that if it's always random, when the first request comes in and it returns `X` builds, it can return `Z` builds when we ask for page `b` based on the previous `total` (it's mostly for pagination purposes).

MiloCasagrande commented on src/routes/projects.js:
```diff
@@ -0,0 +1,175 @@
+  });
+  return;
+});
+
```
You can easily mount here `/:project/lmp/builds` (and all other routes);

MiloCasagrande commented on src/routes/projects.js:
```diff
@@ -0,0 +1,175 @@
+    const name = faker.random.word();
+    arr[idx] = {
+      name,
+      url: `${url}/${bid}/${name}`,
```
We probably have to create a route for the runs as well:
`
/:project/builds/:build/runs/:run
`
Assuming that is the correct url.

MiloCasagrande commented on src/routes/projects.js:
```diff
@@ -0,0 +1,175 @@
+      name,
+      url: `${url}/${bid}/${name}`,
+      status: randomRunStatus(),
+      log_url: `${url}/${bid}/${name}`,
```
Might be worth to have a fake log file to return as well. Some sort of log-formatted output would be best, but even a text file full of `lorem ipsum` will do.

MiloCasagrande commented on src/routes/projects.js:
```diff
@@ -0,0 +1,175 @@
+    data: {
+      build: generateBuildItem({
+        bid: req.params.build,
+        isDetailed: true,
```
What is this field for? If that is `false`, what would change in the output?

marcstreeter commented:
> What is this field for? If that is false, what would change in the output?
this field mimics the api in that the detailed view has more fields than the listing view

marcstreeter commented:
so for whatever reason I am unable to respond to your comments inline here in GitHub so I'll comment on each here:

> Might be worth to have a fake log file to return as well. Some sort of log-formatted output would be best, but even a text file full of lorem ipsum will do.
the log file is really an artifact of the build, there can be any number of artifacts it seems so I went ahead and just did the endpoint and it covers this 

> We probably have to create a route for the runs as well
did that and also did the test one, also I realigned the url's to mimic the real one (not what I was doing before)

> I think we need to define a maximum for total, and make it fixed
fixed it to 500, note only `builds` uses pagination (`runs`, `tests` do not)

> Is annotation always null
for all the ones I was looking at it (as well as `output` was).  I talked to @doanac and got some good info regarding possible `annotation` values (it can be anything but is typically markdown, json strings) - and currently an annotation cannot be set on anything but a promoted build so I put that into the logic as well. As for `output` I just made it a nullable string -- I couldn't find anything that ever had output but I assume it's text (@doanac please steer me if I've guessed wrong)

> Same here as my previous comment on a single faker.random.boolean()
after further review it turns out that the created, completed fields are only there for runs when there are statusEvents so I updated the code for that, and I also made the above request by putting the rest under a single faker.random.boolean

> Add a newline between these, just to make a little bit of clear separation.
put a newline between each function declaration, I'm surprised prettier didn't catch it, I guess it isn't as strict as black is -- I wonder if there is some argument I can give it so it automatically catches that

> For the dates it might be good they are sane: like completed should be in the future compared to started, not completely random.

I made it so the `created` is like a year in the past so that it is always before the completed. I wish `faker` offered some way of making something only a couple of days or months in the past/future. Seems very arbitrary to stick it with a year on certain date functions. I had some other issues with the library because I wanted to generate a random/fake sha. I found a good suggestion in the GitHub page, but the suggested method is not in the current release of the library. Seems like about 2 years ago they were gearing up for version 5.0 but haven't had much movement and their upgrade strategy thus far has been to point people to install it from the master branch - less than ideal. There's a little drama going on over there (100+ PR's that haven't been merged) so I'm not sure if we can depend on that library being updated anytime soon. Not a big deal, still annoying.

MiloCasagrande commented on src/routes/projects.js:
```diff
@@ -0,0 +1,371 @@
+  });
+  return;
+});
+router.get('/:project/builds/:build/', (req, res) => {
```
Add a newline between these as well.

MiloCasagrande commented on src/routes/projects.js:
```diff
@@ -0,0 +1,371 @@
+};
+
+const router = express.Router();
+router.get('/:project/builds/', (req, res) => {
```
We still need to support a route like `/marcstreeter/lmp/builds`: for how express work this route handler will handle just `/marcstreeter/builds`, and if you try to access `/marcstreeter/lmp/builds` it will return 404. The `/lmp` part is considered a path and we don't have any route handling it.

It's mostly the same logic, it's just that the `project` part is `$project/lmp` and not just `$project`.

MiloCasagrande commented:
> put a newline between each function declaration, I'm surprised prettier didn't catch it, I guess it isn't as strict as black is -- I wonder if there is some argument I can give it so it automatically catches that

prettier doesn't do much for that, you will need to have a linter configured to catch those errors.
We'll have to see if eslint has some options to catch missing newlines between function definitions (JS is not that strict compared to Python...), and use prettier and eslint together.

> I made it so the `created` is like a year in the past so that it is always before the completed. I wish `faker` offered some way of making something only a couple of days or months in the past/future. Seems very arbitrary to stick it with a year on certain date functions.

You can always create a simple function to do that based on the actual day (`new Date()`), it's not really that much code if you really want to. What we have now is fine.

>  I had some other issues with the library because I wanted to generate a random/fake sha. I found a good suggestion in the GitHub page, but the suggested method is not in the current release of the library. 

You can also use `crypto.createHash('sha1').update($RANDOM_STUFF).digest('hex')`.

> Seems like about 2 years ago they were gearing up for version 5.0 but haven't had much movement and their upgrade strategy thus far has been to point people to install it from the master branch - less than ideal. There's a little drama going on over there (100+ PR's that haven't been merged) so I'm not sure if we can depend on that library being updated anytime soon. Not a big deal, still annoying.

Yeah, I know. Despite that, it's still the most used and probably the most interesting one.
I mean, the creator/maintainer clearly said that he is looking for help 'cause he doesn't have much time, we might even want to consider helping out, or just fork it and pull in all the PRs that are useful.

Again though, I don't expect to add a lot more work on the faker-api than it needs once we have all the necessary pieces in place.

MiloCasagrande commented on src/routes/projects.js:
```diff
@@ -0,0 +1,175 @@
+    data: {
+      builds: builds,
+      total: builds.length,
+      next: `${url}/builds/?page=1&limit=10`,
```
I was looking at the pagination stuff, and it looks like the `devices` part of the API returns a different pagination data structure. This is what is returned by the devices:
```json5
{
  // Other stuff left out
  "page": 1, "pages": 1, "limit": 20, "total": 1
}
```

The `projects` and I assume also the other endpoints return:
```json5
{
  // Other stuff left out
 "total": 348,
 "next": "https://..."
}
```
We might want to look into it and figure out if we can have the same pagination data as the `devices` endpoint returns, that data structure is way better.

marcstreeter commented:
fixed the routes to accept `/lmp`  (I'll reserve my confusion for why `/` is allowed in a project's name for some other day) and added spaces between function calls

MiloCasagrande commented:
> fixed the routes to accept `/lmp` (I'll reserve my confusion for why `/` is allowed in a project's name for some other day) and added spaces between function calls

`/` is not allowed in a factory/project name, but the API returns the builds for the Linux microPlatform (`lmp`) being built for your factory (we also had `zmp` once), hence the URL is `/$factory-name/lmp/builds`.
